pkg/tsdb/elasticsearch/data_query_processor.go

```diff
@@ -0,0 +1,199 @@
+package elasticsearch
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	es "github.com/grafana/grafana/pkg/tsdb/elasticsearch/client"
+)
+
+// processQuery processes a single query and adds it to the multi-search request builder
+func (e *elasticsearchDataQuery) processQuery(q *Query, ms *es.MultiSearchRequestBuilder, from, to int64) error {
+	err := isQueryWithError(q)
+	if err != nil {
+		return backend.DownstreamError(fmt.Errorf("received invalid query. %w", err))
+	}
+
+	defaultTimeField := e.client.GetConfiguredFields().TimeField
+	b := ms.Search(q.Interval, q.TimeRange)
+	b.Size(0)
+	filters := b.Query().Bool().Filter()
+	filters.AddDateRangeFilter(defaultTimeField, to, from, es.DateFormatEpochMS)
+	filters.AddQueryStringFilter(q.RawQuery, true)
+
+	if isLogsQuery(q) {
+		processLogsQuery(q, b, from, to, defaultTimeField)
+	} else if isDocumentQuery(q) {
+		processDocumentQuery(q, b, from, to, defaultTimeField)
+	} else {
+		// Otherwise, it is a time series query and we process it
+		processTimeSeriesQuery(q, b, from, to, defaultTimeField)
+	}
+
+	return nil
+}
+
+// processLogsQuery processes a logs query and configures the search request accordingly
+func processLogsQuery(q *Query, b *es.SearchRequestBuilder, from, to int64, defaultTimeField string) {
+	metric := q.Metrics[0]
+	sort := es.SortOrderDesc
+	if metric.Settings.Get("sortDirection").MustString() == "asc" {
+		// This is currently used only for log context query
+		sort = es.SortOrderAsc
+	}
+	b.Sort(sort, defaultTimeField, "boolean")
+	b.Sort(sort, "_doc", "")
+	b.AddDocValueField(defaultTimeField)
+	// We need to add timeField as field with standardized time format to not receive
+	// invalid formats that elasticsearch can parse, but our frontend can't (e.g. yyyy_MM_dd_HH_mm_ss)
+	b.AddTimeFieldWithStandardizedFormat(defaultTimeField)
+	b.Size(stringToIntWithDefaultValue(metric.Settings.Get("limit").MustString(), defaultSize))
+	b.AddHighlight()
+
+	// This is currently used only for log context query to get
+	// log lines before and after the selected log line
+	searchAfter := metric.Settings.Get("searchAfter").MustArray()
+	for _, value := range searchAfter {
+		b.AddSearchAfter(value)
+	}
+
+	// For log query, we add a date histogram aggregation
+	aggBuilder := b.Agg()
+	q.BucketAggs = append(q.BucketAggs, &BucketAgg{
+		Type:  dateHistType,
+		Field: defaultTimeField,
+		ID:    "1",
+		Settings: simplejson.NewFromAny(map[string]any{
+			"interval": "auto",
+		}),
+	})
+	bucketAgg := q.BucketAggs[0]
```

Comment:
I wonder what's going on here...? If the slice is non-empty, I think this grabs the old first item, right? Not the one we've just written? I'm not sure what the intent was here. If the array is empty, then there's surely no reason to `append`.